### PR TITLE
executor: fix null div 0 for partition key expression is incorrect 

### LINKED ISSUE #24091: `null div 0` for  partition key expression is incorrect
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
create table t05(a int) partition by hash (a div 0) partitions 10;
insert into t05 values (NULL);
select  null div 0
```

### 2. What did you expect to see? (Required)
```sql
mysql > create table t05 (a int) partition by hash (a div 0) partitions 10;
Query OK, 0 rows affected
Time: 0.088s
mysql > insert into t05 values (NULL);
Query OK, 1 row affected
Time: 0.047s
mysql > select * from t05;
+--------+
| a      |
+--------+
| <null> |
+--------+

```
### 3. What did you see instead (Required)
```sql
mysql > create table t05(a int) partition by hash (a div 0) partitions 10;
Query OK, 0 rows affected
Time: 0.218s
mysql > insert into t05 values (NULL);
(1365, 'Division by 0')
mysql root@172.16.4.131:test> select  null div 0
+------------+
| null div 0 |
+------------+
| <null>     |
+------------+
```
### 4. What is your TiDB version? (Required)

```sql
mysql > select tidb_version()\G
***************************[ 1. row ]***************************
tidb_version() | Release Version: v5.0.0
Edition: Community
Git Commit Hash: bdac0885cd11bdf571aad9353bfc24e13554b91c
Git Branch: heads/refs/tags/v5.0.0
UTC Build Time: 2021-04-06 16:36:29
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: close #24091 

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -784,12 +784,15 @@ func (s *builtinArithmeticIntDivideIntSig) evalIntWithCtx(sctx sessionctx.Contex
 	if isNull || err != nil {
 		return 0, isNull, err
 	}
+	a, isNull, err := s.args[0].EvalInt(sctx, row)
+	if b == 0 && isNull {
```
Comment (mmyj) [inline expression/builtin_arithmetic.go:788]: Should we check err before L788?

**Review by lzmhhh123 — APPROVED** [review]

**Pull request merged**